docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,673 @@
+   * Returns an instantiated module
+   */
+  public async executeEntrypoint(url: string): Promise<Record<string, any>>
+  /**
+   * Other ModuleRunner methods...
+   */
+```
+
+The Node module runner instance is accessible through `server.nodeModuleRunner`. It isn't part of its associated environment instance because, as we explained before, other environments' module runners could live in a different runtime (for example for the browser, a module runner in a worker thread as used in Vitest, or an edge runtime like workerd). The communication between `server.nodeModuleRunner` and `server.environment('node')` is implemented through direct function calls. Given a Vite server configured in middleware mode as described by the [SSR setup guide](#setting-up-the-dev-server), let's implement the SSR middleware using the environment API. Error handling is omitted.
+
+```js
+app.use('*', async (req, res, next) => {
+  const url = req.originalUrl
+
+  // 1. Read index.html
+  let template = fs.readFileSync(path.resolve(__dirname, 'index.html'), 'utf-8')
+
+  // 2. Apply Vite HTML transforms. This injects the Vite HMR client,
+  //    and also applies HTML transforms from Vite plugins, e.g. global
+  //    preambles from @vitejs/plugin-react
+  template = await server.transformIndexHtml(url, template)
+
+  // 3. Load the server entry. executeEntryPoint(url) automatically transforms
+  //    ESM source code to be usable in Node.js! There is no bundling
+  //    required, and provides full HMR support.
+  const { render } = await server.nodeModuleRunner.executeEntryPoint(
+    '/src/entry-server.js',
+  )
+
+  // 4. render the app HTML. This assumes entry-server.js's exported
+  //     `render` function calls appropriate framework SSR APIs,
+  //    e.g. ReactDOMServer.renderToString()
+  const appHtml = await render(url)
+
+  // 5. Inject the app-rendered HTML into the template.
+  const html = template.replace(`<!--ssr-outlet-->`, appHtml)
+
+  // 6. Send the rendered HTML back.
+  res.status(200).set({ 'Content-Type': 'text/html' }).end(html)
+})
+```
+
+## Environment agnostic SSR
+
+One of the objectives of this proposal is to allow users to swap Vite's default SSR environment. Out-of-the-box, Vite will use a node environment running on the same process as the server. But a user may want to swap it for a workerd environment.
+
+Vite would expose a `server.ssrEnvironment` so all frameworks can use it and allow them to define what environment type should be used for it. Ideally, instead of using `server.nodeModuleRunner`, the example above would be written as:
+
+```js
+// 3. Load the server entry, with full HMR support.
+const { render } = await server.ssrEnvironment.run('/src/entry-server.js')
+
+// 4. render the app HTML.
+const appHtml = await render(url)
+```
+
+`run(url)` would return RPC wrappers for `entry-server` exported functions, so that this code is compatible with both node and workerd environments.
```

Comment:
This is very interesting. I'm leaning towards having good error messages here, rather than restricting what the user can do with the Node Environment. At least for Astro and SvelteKit I know they will be more than happy to review and fix their code as soon as they see they can get workerd SSR in dev by doing so. And we may be able to work with the authors of popular plugins too.

But you have a good point. Maybe there is a way to warn users and force them to disable the warning for example? My only worry is that there may be cases where Vite is used to develop tooling and not a framework that is only intended to work in Node and I don't think we should restrict them in that case.

We could have a "safe" mode where we go through serialization even when using the node environment to make sure it will work everywhere.

Reply:
I agree that `run` function should always enforce serializable data. For loading Node-SSR code, frameworks can always use `nodeRunner` or something maybe 🤷🏻 I am also ok with exposing an option to enable enforcement.

Another point to consider is if we even want to expose the module. Currently, a lot of frameworks use `ssrLoadModule` and call methods freely, but with the new API, they would _have_ to refactor their code, so maybe it's better to limit this function to simply running the module itself, and not return any RPC layer. On the other hand, I see how it can be useful to expose some methods knowing the limitations to establish routes for example:

```js
const { startRouter } = await environment.run('/entry.js')
await startRouter(['/api', '/api/puppies'])
```

The alternative would be to run separate modules:

```js
await environment.run('/entry.js?routes=/api,/api/puppies')
// calls "startRouter" inside the module
```

Reply:
Coming from the comment https://github.com/vitejs/vite/pull/16089#discussion_r1515592481 and especially for Workerd integration, I think loading "entrypoint file" is a part of Miniflare environment initialization, so the API such as `run: (entyFile: string) => Promise<anyModule>` didn't seem easy to provide based on the current approach used by experiments so far. 

This is not what I'm familiar with, but maybe durable objects would be a good fit to implement such "proxy module"?

Reply:
Added a clarification after internal discussions (it still need a better explanation):  https://github.com/vitejs/vite/pull/16089/commits/650106656b0aa8dcb45fc8a277a5b5c796f5dcf4

We are thinking about separating the RPC layer into a feature we could deliver independently. If there is demand and we find a good way to do it, I think it would be very helpful to make environments easier to use but it isn't strictly needed.

So it is being proposed that the first MVP will only have `environment.run(url): void`.

And we could later on (if we find a good way to do it), a `environment.import(url): RPC exports` (or `load`?). According to @IgorMinar they are seeing that a lot of frameworks end up re-implementing a RPC scheme, so there could be value in sharing it at the Vite core level. And I like that it makes the node environment and the workerd environment a lot easier to swap, given that we help users avoid non-serializable data by having opt-in or opt-out guards against that. @sheremet-va mention yesterday that if users want to use non-serializable data, they could directly use the node runner:
```js
// RPC export, even for the node environment, can't be opt-out, we ensure that
// the environment can be swapped by any other environment that is able to 
// implement the RPC `import` function.
const { render } = environment.import(url)

// Equivalent to `ssrLoadModule`, there is no restriction here
const { render } = moduleRunner.import(url)

// The node module runner is exposed as `server.nodeModuleRunner` and is part of 
// Vite's default toolbox when you create a dev server.
const { render } = server.nodeModuleRunner.import(url)

// But you won't be using it for SSR, as we want to allow swappable environments,
// so you'll use the exposed `server.ssrEnvironment` that frameworks define that works
// on all environments that implemented RPC exports for import
const { render } = server.ssrEnvironment.import(url)

// We should only recommend this for SSR if we find a good way to make RPC work
// in workerd. Until then, we would propose using `environment.run(url): void`
server.ssrEnvironment.run(url)
// and so on... 
```

Reply:
If we are first going only with `run`, I guess we need to expose something like `transformIndexHtml` on ModuleRunner.
```ts
// in node world

server.ssrEnvironment.run('/entry.js')
```
```ts
// entry.js running in other runtime (e.g. workerd)

const templatePath = '/index.html'

export default async (req: Request) => {
  const template = await moduleRunner.transformIndexHtml(req.url, templatePath) // this function

  const renderer = await moduleRunner.executeEntrypoint()
  const appHtml = renderer(req.url)

  const html = template.replace(`<!--ssr-outlet-->`, appHtml)

  return new Response(html, { headers: { "content-type": "text/html;charset=UTF-8" } })
}
```
Also I guess going this way is not compatible with plugins using `configureServer` hook to inject middlewares unless they make a middleware that proxies requests from Vite server to the other server.

Reply:
We can brainstorm some kind of a solution. Maybe returning a string is fine? Or do we require a Response? Should we push standard Web APIs here? We can use this opportunity to refine the SSR API that relies on running code on another server, and keep `ssrLoadModule` as a separate thing just for running code on the server

Reply:
> I guess we need to expose something like `transformIndexHtml` on ModuleRunner.

Also, can't you just use it as a virtual module? 🤔 It should just return a string, right? So in the workerd it's something like this:

```ts
const html = import.meta.env.DEV ? await import(`virtual:ssr-html/${req.url}`).default : await import('/index.html').default
return new Response(html.replace('<!-- ssr -->', reactRenderer.renderToString(<App>)))
```

Reply:
Ah, that's true.